cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -468,13 +485,15 @@ class DatasetWriter::DatasetWriterImpl : public util::AsyncDestroyable {
   }
 
   Future<> DoWriteRecordBatch(std::shared_ptr<RecordBatch> batch,
-                              const std::string& directory) {
+                              const std::string& directory, const std::string& prefix) {
     ARROW_ASSIGN_OR_RAISE(
         auto dir_queue_itr,
         ::arrow::internal::GetOrInsertGenerated(
-            &directory_queues_, directory, [this, &batch](const std::string& dir) {
-              return DatasetWriterDirectoryQueue::Make(
-                  &task_group_, write_options_, &writer_state_, batch->schema(), dir);
+            &directory_queues_, directory + prefix,
+            [this, &batch, &directory, &prefix](const std::string& dir) {
```

Comment:
```suggestion
            [this, &batch, &directory, &prefix](const std::string& key) {
```
A small change that may help future readers understand why we need to capture `directory` instead of using `dir`.

Reply:
Made the change.